core/src/stream_id.rs

```diff
@@ -0,0 +1,229 @@
+use crate::{Bytes, EncodedLen};
+use cid::multibase::{self, Base};
+pub use cid::Cid;
+use std::fmt::Formatter;
+use std::io::Write;
+
+use int_enum::IntEnum;
+use serde::{Deserialize, Deserializer, Serialize};
+use unsigned_varint::{decode, encode};
+
+/// Types of possible stream id's
+#[repr(u64)]
+#[derive(Copy, Clone, Debug, Eq, IntEnum, PartialEq)]
+pub enum StreamIdType {
+    /// A model stream id
+    Model = 2,
+    /// A document stream id
+    Document = 3,
+    /// Unloadable. This is also used for the parent stream id of all models
+    Unloadable = 4,
+}
+
+impl Serialize for StreamIdType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u64(self.int_value())
+    }
+}
+
+/// A stream id, which is a cid with a type
+#[derive(Debug, Eq, PartialEq)]
+pub struct StreamId {
+    /// The type of the stream
+    pub r#type: StreamIdType,
+    /// Cid of the stream
+    pub cid: Cid,
+    /// Optional commit for this stream
+    pub commit: Option<Cid>,
```

Comment:
Do we need to introduce "commitIds" yet (or at all)? 

There is a painpoint around how commitIds are specified currently because they kind of break the multiformat nature of streamIds.

Me and @AaronGoldman talked about introducing a new way to reference a specific event inside of a streamId. This approach would allow streamIds to always be a real multiformat:
https://github.com/AaronGoldman/CIP/blob/stream-sets-synchronization/CIPs/CIP-124/cip-124.md#eventid-sort-order

Reply:
I mainly have it there to make sure we can read cid's with commit id's specified. If we don't think we'll have stream id's with commits, I'd gladly remove it.